### ISSUE TITLE
Fingered Portamento causing infinite playback.

### ISSUE THREAD

**Issue by ra-om** [surge-synthesizer/surge#6287]
**Bug Description:**
The last note is often played to infinity after several notes of fingered portamento, and the plugin preset has to be reset manually for it to stop.

**Surge XT Version**
XT 1.0.1

* Version:
* Plugin Type: VST3
* Bitness: 64-bit

**Reproduction Steps:**
It is more likely to happen when chaining 3 or more notes with FP, but can happen with less. Attached is a reaper project with a MIDI bit that triggers it 100% of the time, with that preset at least.

**Expected Behavior:**
The last note to end on NoteOff.

**Computer Information (please complete the following!):**
 - OS: Debian 64-bit
 - Host: Reaper
 - Version: 6.54

Preset is Kinsey Dulcet's "Holy Emoji", under Sequences.

[DulcetBugReport.zip](https://github.com/surge-synthesizer/surge/files/8886767/DulcetBugReport.zip)

**Comment by mkruselj** [comment]
Can you try the latest nightly?

**Comment by baconpaul** [comment]
So I just tried this at head

1. The bug is real. The voice never goes off
2. Nothing to do with the patch. Change to init saw and it happens
3. Those midi notes are *loooow*. I had to go up 3 octaves to hear it at all

I wonder if there is a midi note 0 problem somewhere

**Comment by ra-om** [comment]
It turns out the MIDI example I gave had mixed channels, first 3 notes on channel 5 and the rest on channel 1. I am not able to reproduce infinite playback when all notes are of the same channel, but encountered some weird behaviors(apart from the sound of the phrasing which changes drastically):



[SameAndMixedChan.zip](https://github.com/surge-synthesizer/surge/files/8893395/SameAndMixedChan.zip)


I'll open a separate issue for the other behaviors.

**Comment by ultrasound1372** [comment]
Could this also be related to some control surfaces hacking a note-off by sending a note-on with a matching pitch and 0 velocity?

**Comment by baconpaul** [comment]
Ahh ok @ra-om I took a look

the dulcet patch holy emoji is a dual scene patch. In this mode, certain midi channels map to certain scenes.

I think what is happening is by mixing the midi channels you are turning notes on but not off on one of the scenes. I scoured the code and I really don't see a bug (even with velocity 0) for midi-consistent streams of data.

Did you ever open an issue for other behaviors? I"m going to close this one after investigating - I don't *think* there's an open issue which requires code but please reopen with comment if i"m wrong.

**Comment by ra-om** [comment]
The other behaviors were related to channel 2 and 3 mapping to scenes A and B, which turns out is in the manual, so I closed it soon after.

The thing is @baconpaul, that the bug happens even with Init Saw in single scene mode, and the notes in the MIDI snippet are channels 5 and 1; while the manual only mentions channel 2 and 3 for Scenes.

I think there is still a bug here, please investigate further. How do I reopen an existing issue?

**Comment by baconpaul** [comment]
I re-opened it

Are the note on and off on the same channel? Can you give me an example of the midi stream that breaks init saw?

Thanks

**Comment by ra-om** [comment]
![image](https://user-images.githubusercontent.com/12862942/183267395-fbc206a8-227a-4d48-bff8-24d1fc46dec8.png)
[infinoteBug.zip](https://github.com/surge-synthesizer/surge/files/9275824/infinoteBug.zip)

Could the lack of an "all notes off" channel 5 event be the cause?

It also happens in Surge 1.9 BTW

**Comment by baconpaul** [comment]
Very useful.  Thank you

But I've not been able to make this midi file stick a note. here's what I did

1. I opened reaper mac and loaded the surge vst3
2. I dragged that midi file on, loaded it as multi-channel single track, and played it, no problem

So then also tested the midi in bitwig and logic pro with the clap and au and no problem either.

in no mode (poly, monofp, etc...) does that give me stuck notes.

So can you give me some more instructions on how to see the stuck note? Or maybe send an RPP file zipped that does it?

I'm sorry this is so hard to find but i can't reproduce it so i'm not sure how to help.

You're sure you don't have any midi transforms in line or anything right? The fact that it happens in Surge 1.9 (which is for most purposes a very different synth) makes me think it is something odd!

**Comment by ra-om** [comment]
[infinoteBug_1.zip](https://github.com/surge-synthesizer/surge/files/9276009/infinoteBug_1.zip)

I did the same steps you followed and still get the stuck note only with the Single Trigger modes, both with and without FP.

Attached project is a new project with the MIDI file imported.

**Comment by baconpaul** [comment]
Thanks.

**Comment by baconpaul** [comment]
<img width="956" alt="Screen Shot 2022-08-07 at 10 12 02 AM" src="https://user-images.githubusercontent.com/13387561/183294901-8ac38db0-06b7-4e05-80f6-d9421aba3039.png">

OK just this minimal case sticks. I should be able to repro that in the test harness. Thank you.

**Comment by baconpaul** [comment]
Yeah cool

```cpp
TEST_CASE("Mono ST Across Channels", "[midi]")
{
    for (auto secondChannel : {1, 5})
    {
        DYNAMIC_SECTION("On on 1, Off on " << secondChannel)
        {
            auto surge = Surge::Headless::createSurge(44100);
            REQUIRE(surge);
            surge->storage.getPatch().scene[0].polymode.val.i = pm_mono_st;

            for (int i = 0; i < 10; ++i)
                surge->process();

            surge->playNote(secondChannel - 1, 60, 127, 0);
            for (int i = 0; i < 50; ++i)
                surge->process();
            REQUIRE(surge->voices[0].size() == 1);

            surge->playNote(0, 65, 127, 0);
            for (int i = 0; i < 50; ++i)
                surge->process();
            REQUIRE(surge->voices[0].size() == 1);

            surge->releaseNote(secondChannel - 1, 60, 0);
            for (int i = 0; i < 50; ++i)
                surge->process();
            REQUIRE(surge->voices[0].size() == 1);
            
            surge->releaseNote(0, 65, 0);
            for (int i = 0; i < 50; ++i)
                surge->process();
            REQUIRE(surge->voices[0].size() == 0);
        }
    }
}
```

that test fails on the second case. Great.

**Comment by baconpaul** [comment]
OK I'll merge that test (with the failure condition knocked out) for now

The problem is that (roughly) https://github.com/surge-synthesizer/surge/blob/eeaf67675064db6f174fed20f2021c6a555a1d70/src/common/SurgeSynthesizer.cpp#L925 that line checks the channel of the inbound note. We also need to check all the other channels which this scene *could* play to look for a voice so that loop has to change in a way which is a bit subtle and scary.

Good chance that the test I merged would also fail in the reset-from-zero case even in FP also so want to consider all of that.

Luckily the fix from a user perspective is easy. If you have a coherent mono voice, use a coherent midi channel. So I don't want to rush a fix for this but it's a real bug-like-feature of our channel handling indeed. Thank you for the help in identifying it.

**Comment by baconpaul** [comment]
Yeah the new methods all fail also. OK updated the test so when we tackle this we have the test there.

**Comment by mkruselj** [comment]
Possible solution: #4657.

Add that option. When that option is disabled convert ANY incoming MIDI to channel 1, except when we're in MPE mode.

**Comment by baconpaul** [comment]
Yeah I was thinking in non ch and non dual mode without MPE remap midi etc…. 

It’s not clear that fixing the voice manager vs the harness is the right answer though I agree

**Comment by mkruselj** [comment]
I think Surge is not doing anything wrong here. If a note on comes through some channel it is not Surge's responsibility to convert note off coming from *another* channel to the other channel in order to stop the note.

**Comment by baconpaul** [comment]
So the pattern which sticks is

on 5 60
On 1 67
Off 5 60
Off 1 67

that shouldn’t stick

**Comment by baconpaul** [comment]
But yes mono modes ignoring channels which are not in the mask is another solution I considered but that would break patches.

**Comment by mkruselj** [comment]
Maybe a sort of thing to break for XT2 then.

**Comment by baconpaul** [comment]
yeah maybe

like i said we know the problem and there's a user solution. so i don't want to rebuild the voice system in non compatible ways. This one may stay open now diagnosed. Thinking.

**Comment by ra-om** [comment]
> Thanks.

You're welcome, thanks for all the great work on this synth
